### PR TITLE
Issue 663

### PR DESCRIPTION
**Pull request by tiago-loureiro**
This PR (temporarily) introduces the concept of _currency_ in our team events, which get propagated `brig -> galley -> billing` during team creation/activation.

Some considerations:
 * This currency is only propagated in case of team activation during registration, i.e., if there is a pre-verified e-mail address
 * We allow all [ISO_4217 currencies](https://en.wikipedia.org/wiki/ISO_4217) to be specified during creation @dkovacevic 
 * The encoding of the currency in the protobuf is done as an optional string, which is not ideal, but prevents having to map all currencies to a protobuf enum (or a subset, requiring then new builds for any added currency). If someone some better ideas here, please let me know :)
 * The intend is also to revert this PR once we are able to set currencies _after_ team activation

Note that some team tests were also rewritten as [this test](https://github.com/wireapp/wire-server/blob/develop/services/brig/test/integration/API/Team.hs#L62) was failing, due to the fact that the user got activated before the team creation (possible with the usage of internal APIs).

### CODE REVIEW

File: libs/brig-types/stack.yaml [hunk 1]
```diff
@@ -10,5 +10,6 @@ flags:
     protobuf: True
     arbitrary: True
 
-extra-deps: []
+extra-deps:
+- currency-codes-2.0.0.0
```
Comment (elverkilde) [inline libs/brig-types/stack.yaml]: why is this needed?

Comment (tiago-loureiro) [reply libs/brig-types/stack.yaml]: No `currency-codes` on [stackage](https://www.stackage.org/package/currency-codes) unfortunately

Comment (elverkilde) [reply libs/brig-types/stack.yaml]: as far as I understand, this will end up in Stripe at some point? Would it be worth looking into `stripe-core`? it seems to enumerate the currencies supported by Stripe:
https://www.stackage.org/haddock/lts-10.2/stripe-core-2.2.3/Web-Stripe-InvoiceItem.html#t:Currency

Comment (dkovacevic) [reply libs/brig-types/stack.yaml]: Stripe works slightly different. Customer is created independently from the currency. Then `ibis` creates new _Subscription_, for this new _Customer_, by choosing between various _Plans_ ( `monthly_eur`, `monthly_usd`, `monthly_chf`).  Those _Plans_ were previously created in Stripe Dashboard.

Comment (tiago-loureiro) [reply libs/brig-types/stack.yaml]: @dkovacevic I think @elverkilde's point was just that instead of using [currency-codes](https://hackage.haskell.org/package/currency-codes) to manage currencies we could as well just use [stripe-core](https://www.stackage.org/haddock/lts-10.2/stripe-core-2.2.3/Web-Stripe-InvoiceItem.html#t:Currency) which has sort of the same functionality (with the added value that it already rejects some currencies that stripe will anyway not accept)

**Comment by elverkilde** [comment]
overall LGTM

File: services/brig/src/Brig/API/User.hs [hunk 2]
```diff
@@ -467,48 +468,49 @@ activate :: ActivationTarget
          -> ActivationCode
          -> Maybe UserId -- ^ The user for whom to activate the key.
          -> ExceptT ActivationError AppIO ActivationResult
-activate tgt code usr = do
-    key <- mkActivationKey tgt
-    activateKey key code usr
+activate tgt code usr = activate' tgt code usr Nothing
 
-activateKey :: ActivationKey
+activate' :: ActivationTarget
```
[the remark this answers is not in the document]
Comment (tiago-loureiro) [reply services/brig/src/Brig/API/User.hs]: 👍

File: libs/brig-types/src/Brig/Types/User.hs [hunk 3]
```diff
@@ -275,7 +276,26 @@ newtype InvitationCode = InvitationCode
...
+        return $ BindingNewUserTeam t c
+    parseJSON _ = fail "parseJSON BindingNewUserTeam: must be an object"
+
+instance ToJSON BindingNewUserTeam where
+    toJSON (BindingNewUserTeam t c) = do
+        let (Object t') = toJSON t
+         in object $ "currency" .= c
+                   # HashMap.toList t'
+
+newtype NewUserTeam = NewUserTeam { nuTeam :: Either InvitationCode BindingNewUserTeam}
```
[the remark this answers is not in the document]
Comment (tiago-loureiro) [reply libs/brig-types/src/Brig/Types/User.hs]: Hmmmm, true. How about something like
```
data NewUserTeam = NewTeamMember  !InvitationCode
                 | NewTeamCreator !BindingNewUserTeam
```

[the remark this answers is not in the document]
Comment (elverkilde) [reply libs/brig-types/src/Brig/Types/User.hs]: Yes, something like that. The naming is also a bit confusing, if this is a user I would expect it to end in User, i.e. `NewTeamUser`

[the remark this answers is not in the document]
Comment (tiago-loureiro) [reply libs/brig-types/src/Brig/Types/User.hs]: 👍

**Comment by tiago-loureiro** [comment]
Will merge this in only after https://github.com/wireapp/wire-server/pull/163, https://github.com/wireapp/wire-server/pull/162 and https://github.com/wireapp/wire-server/pull/161 so that we can make a release to master before merging this one in
